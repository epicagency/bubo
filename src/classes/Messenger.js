import * as locales from '../locales';
import { parseTpl } from '../helpers/utils';

/**
 * Handle messages
 *
 * @class Messenger
 */
class Messenger {
  /**
   * Creates an instance of Messenger.
   * @param {string} [locale='en'] default locale is english
   *
   * @memberOf Messenger
   */
  constructor(locale = 'en') {
    this.locale = locale;
    this._messages = this.getMessages();
  }

  /**
   * Get locale value
   * @readonly
   *
   * @memberOf Messenger
   */
  get locale() {
    return this._locale;
  }

  /**
   * Set new locale value
   * @param {string} locale available locale
   *
   * @memberOf Messenger
   */
  set locale(locale) {
    if (Messenger._hasLocale(locale)) {
      this._locale = locale;
      this._messages = this.getMessages();
    }
  }

  /**
   * Check if locale exists
   *
   * @static
   * @param {string} locale locale to check
   * @returns {boolean} returns true if locale exists
   *
   * @memberOf Messenger
   */
  static _hasLocale(locale) {
    if (!locales[locale]) {
      throw new Error(`ðŸ¦‰ locale "${locale}" is not available!`);
    }

    return true;
  }

  // !TODO addLocale()

  /**
   * Get messages
   *
   * @returns {Object} messages
   *
   * @memberOf Messenger
   */
  getMessages() {
    return locales[this.locale];
  }

  /**
   * Get global text message depending on status
   *
   * @param {string} status status
   * @returns {string} text message
   */
  getText(status) {
    return this._messages.status[status];
  }

  /**
   * Get error message
   *
   * @param {string} rule rule name
   * @param {string} [label=null] dynamic label
   * @param {any} [value=null] dynamic value(s)
   * @returns {string} error message
   */
  getError(rule, label = null, ...values) {
    const errorMessages = this._messages.errors;
    const strings = {
      label: label || errorMessages.defaultLabel,
    };

    values.forEach((val, i) => {
      strings[`value${i}`] = val;
    });

    return parseTpl(errorMessages.rules[rule], strings);
  }
}

export default Messenger;
