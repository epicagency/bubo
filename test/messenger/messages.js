import test from 'ava';
import Messenger from '../../src/classes/Messenger';

test('Success text', (t) => {
  const msg = new Messenger();

  t.snapshot(msg.getText('success'));
});

test('Error text', (t) => {
  const msg = new Messenger();

  t.snapshot(msg.getText('error'));
});

test('Default error messages', (t) => {
  const msg = new Messenger();

  t.regex(
    msg.getError('required'),
    /^This field.+required.*/
  );

  t.regex(
    msg.getError('required', 'Test'),
    /^Test.+required.*/
  );

  t.regex(
    msg.getError('min', 'Test', 2), // eslint-disable-line
    /^Test.+2.*/
  );

  t.regex(
    msg.getError('minmax', 'Test', 2, 3), // eslint-disable-line
    /^Test.+2 and 3.*/
  );
});

test('Message changes with locale', (t) => {
  const msg = new Messenger();

  msg.locale = 'fr';
  t.regex(
    msg.getError('required'),
    /^Ce champs.+requis.*/
  );
});
