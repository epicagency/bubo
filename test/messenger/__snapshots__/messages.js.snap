// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Error text 1`] = `"Please correct errors."`;

exports[`Success text 1`] = `"The form is completed correctly."`;
